### ISSUE TITLE
casm update takes about an hour to update a 108 atom supercell

### ISSUE THREAD

**Issue by mxgrabowski** [prisms-center/CASMcode#54]
I am trying to set up a data base for a ternary cluster expansion and I set up the Ni-Re-Va configurations by hand and calculated them using vasp. So, I had to import them correctly to my casm project using `casm import -i --pos POSCAR`. In the beginning the import seemed to take quite a while, so the workaround was to create the ideal structure for the POSCAR (My POSCAR files were pre-relaxed). Then the import was quick and I could created the corresponding `properties.calc.json` files. However, these have to be updated with `casm update` which takes also long time as these large supercells (107 atoms with 1 vacancy) have to be mapped. I was able to lower the time for an update of a single configuration to ~55min with `casm update --max-vol-change 0.01 --max-va-frac 0.0093 --min-va-frac 0.0092`. This is still a long time to update a single configuration, because I have 39 of them. Are there possibilites to fasten up the update, because I already got the supercell and the configuration via `casm import`.

Thank you very much for the help!

**Comment by jcthomas** [comment]
This is a problem that pops up from time to time. The difficulty arises because CASM makes no assumptions about which configuration a relaxed structure should map onto, so it performs an exhaustive search of all possibilities. 

The next major release of CASM should include some more refined options to restrict this search.

Until then, since you have the ideal structures, you can copy your ideal POSCARs to CONTCAR, and then generate properties.calc.json, with the ideal structure's coordinates. Then, run 

`casm import -i --pos path/to/properties.calc.json -d`

(or you can use the bach import option to import multiple at once)

which will import the data with the ideal configs. This should be just as fast as your previous 'ideal' import. Note, however, that, with this workflow, CASM will be unable to detect any mechanically-unstable configurations. This is unlikely to be an issue if your configurations are in large supercells and are vacancy-dilute, but it is important to be aware of.

**Comment by mxgrabowski** [comment]
Thanks for your response.

First, I tried to update all calculations with `casm update --max-vol-change 0.01 --max-va-frac 0.0093 --min-va-frac 0.0092`, but after 4 days casm was able to only work on 3 configurations, so I interrupted it. Now, I am trying your approach. But how do I generete `properties.calc.json` with the ideal structure's coordinates? I copied the ideal POSCAR to CONTCAR, but using `vasp.relax.report ./` and `casm-calc --report` seems to only use `vasprun.xml` to write out the structure. So I am not able to create a `properties.calc.json` with ideal positions. 
`casm import -i --pos path/to/properteis.calc.json -d` also takes a lot of time.

Is there a possibility to read in the POSCAR/CONTCAR to create the `properties.calc.json`?

**Comment by jcthomas** [comment]
I forgot that everything comes from `vasprun.xml`.  I made a branch on my repo, which you can find at https://github.com/jcthomas/CASMcode/tree/0.2.X_faster_update. It allows you to force relaxed structures to map exactly onto their starting configs via `casm update --strict`. This should enable you to finish your update quite quickly, at the cost of not being able to tell if any configurations are mechanically unstable.

Try that out, and if it does what you need, I'll open a pull request to include it in the main repo.

If you think some of your configurations could be mechanically unstable, we might consider adding an option that would restrict the supercells that are considered during the mapping search.

**Comment by mxgrabowski** [comment]
I tried using your branch on your repo. I was not able to install your branch, so I checked what are the change you made in `ConfigMapping.cc`. I adopted the changes in my branch (master; v0.2.X) and I was able to compile it. I tried `casm update --strict`, but it still seems to take a large amount of time, so I searched through the code. 

The changes you made are in `ConfigMapping.cc` in lines 191+. As I understand it, you want to skip this if condition. However, using `--strict`, we never arrive at this condition. This condition is in the `import_structure_occupation function` from `update.cc`. I looked into `update.cc` and in lines 196+ we are calling `import_structure` and not `import_structure_occupation`, if `--strict` is set to true. I checked the `import_structure` in `ConfigMapping.cc` and narrowed it down where the code is hanging while updating the structures. We're getting to lines 326+ and call the function `struc_to_configdof`. Looking at this function, we have a if condition in line 389 and the bool in this statement is `false`, so we continue in line 400 as `!valid_mapping` is `true` and we call the function `struc_to_configdof_deformed`. This is where `casm update --strict` is hanging and I wasn't able to identify where the code is taking much more time in this function.

**Comment by jcthomas** [comment]
Sorry, that use of `import_structure` in `update.cc` is incorrect.  Only `import_structure_occupation` should be used in the current version.  I updated my branch, and you can check out the changes. 

`struc_to_configdof_deformed` is exactly what I was trying to avoid with the patch. It performs a rigorous search over all compatible supercells to find the one that is the best mapping for the structure under consideration.

To your other point, my branch (jcthomas/CASMcode/0.2.X_faster_update) is only slightly different from prisms-center/CASMcode/0.2.X. What is the issue you are having with compilation?

**Comment by mxgrabowski** [comment]
I applied your changes in the `update.cc` file and with the `casm update --strict` command I was able to quickly update my large super cells!

Regarding the compilation. I think made a mistake getting this branch from github. Sanjeev helped me getting the correct branch with the up-to-date files. I compiled the new version and this one seems to work fine.

Although, I have an older issue loading the `libcasm.so` file, but I will open a new issue with this problem.
